core/src/main/java/com/linecorp/armeria/common/stream/AbstractMulticastStreamFactory.java

```diff
@@ -0,0 +1,574 @@
+                int upstreamOffset = this.upstreamOffset;
+                if (upstreamCompleted && offsetOfSubscriber == upstreamOffset) {
+                    if (DownstreamSubscription.writeFlagUpdater.compareAndSet(downstream, 0, 1)) {
+                        completeDownstream(downstream, upstreamCause);
+                    }
+                    return;
+                }
+
+                if (downstream.cancelled && offsetOfSubscriber == upstreamOffset) {
+                    if (DownstreamSubscription.writeFlagUpdater.compareAndSet(downstream, 0, 1)) {
+                        completeDownstream(downstream, CancelledSubscriptionException.get());
+                    }
+                    downstreamSubscriptions.remove(downstream);
+                    return;
+                }
+
+                if (offsetOfSubscriber == upstreamOffset) {
+                    break;
+                }
+
+                if (!publishData(downstream)) {
+                    break;
+                }
+            }
+        }
+
+        private void completeDownstream(DownstreamSubscription downstream, Throwable cause) {
+            if (cause == null) {
+                try {
+                    downstream.subscriber().onComplete();
+                } finally {
+                    @SuppressWarnings("unchecked")
+                    final CompletableFuture<Void> f =
+                            (CompletableFuture<Void>) downstream.streamMessage().closeFuture();
+                    f.complete(null);
+                }
+            } else {
+                try {
+                    if (!(cause instanceof CancelledSubscriptionException)) {
+                        downstream.subscriber().onError(cause);
+                    }
+                } finally {
+                    downstream.streamMessage().closeFuture().completeExceptionally(cause);
+                }
+            }
+        }
+
+        private boolean publishData(DownstreamSubscription downstream) {
+            for (;;) {
+                final long demand = downstream.demand;
+                if (demand == 0) {
+                    break;
+                }
+
+                if (demand == Long.MAX_VALUE ||
+                    DownstreamSubscription.demandUpdater.compareAndSet(downstream, demand, demand - 1)) {
+                    final Subscriber<Object> subscriber = downstream.subscriber();
+                    if (DownstreamSubscription.writeFlagUpdater.compareAndSet(downstream, 0, 1)) {
+                        final T content = contentList.get(downstream.offset++);
+                        if (content instanceof ByteBufHttpData) {
+                            ByteBuf buf = ((ByteBufHttpData) content).buf();
+                            if (!downstream.withPooledObjects()) {
+                                subscriber.onNext(HttpData.of(ByteBufUtil.getBytes(buf)));
+                            } else {
+                                buf.retain();
```

Comment:
Don't we need to make a new `ByteBufHttpData` instance with `ByteBuf.retainedDuplicate()`?